File: scripts/release.mjs
#!/usr/bin/env node
// Cut an entviz-js release: bump version, commit, push, tag, push tag.
//
// This is a HUMAN-run script. AGENTS.md reserves pushes to main and tags for
// humans; AI agents must not run this.
//
// The two workspace packages (@entviz/core, @entviz/react) are versioned in
// LOCKSTEP — every release bumps both to the same X.Y.Z, and @entviz/react's
// dependency on @entviz/core is pinned to that exact version. @entviz/core's
// version is the single source of truth that this script reads and bumps.
//
// Usage:
//   node scripts/release.mjs                       # patch bump, default message
//   node scripts/release.mjs -m "add foo"          # patch bump, custom message
//   node scripts/release.mjs --minor -m "new API"  # minor bump
//   node scripts/release.mjs --major -m "rewrite"  # major bump
//   node scripts/release.mjs --set 0.5.0 -m "..."  # set an explicit version
//                                                  #   (must be > current, and
//                                                  #    may not jump the major
//                                                  #    by >1 without
//                                                  #    --allow-major-jump)
//
// After the tag reaches GitHub, .github/workflows/release.yml verifies the tag
// matches the manifest version, runs the tests, and publishes to npm.

import { readFileSync, writeFileSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { dirname, join, relative } from "node:path";
import { createInterface } from "node:readline";

const REPO_ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const CORE_PKG = join(REPO_ROOT, "packages/core/package.json");
const REACT_PKG = join(REPO_ROOT, "packages/react/package.json");
const CORE_NAME = "@entviz/core";

function die(msg) {
  console.error(msg);
  process.exit(1);
}

function run(cmd, args, { capture = false } = {}) {
  return execFileSync(cmd, args, {
    cwd: REPO_ROOT,
    encoding: "utf8",
    stdio: capture ? ["ignore", "pipe", "inherit"] : "inherit",
  });
}

function get(cmd, args) {
  return run(cmd, args, { capture: true }).trim();
}

function readJson(path) {
  return JSON.parse(readFileSync(path, "utf8"));
}

function currentVersion() {
  const v = readJson(CORE_PKG).version;
  if (!/^\d+\.\d+\.\d+$/.test(v ?? "")) {
    die(`Could not read a valid X.Y.Z version from ${relative(REPO_ROOT, CORE_PKG)} (got ${v}).`);
  }
  return v;
}

function bump(version, part) {
  const [major, minor, patch] = version.split(".").map(Number);
  if (part === "major") return `${major + 1}.0.0`;
  if (part === "minor") return `${major}.${minor + 1}.0`;
  return `${major}.${minor}.${patch + 1}`;
}

function parseExplicit(value, current, { allowMajorJump }) {
  if (!/^\d+\.\d+\.\d+$/.test(value)) die(`--set expects X.Y.Z (got ${value}).`);
  const [n, c] = [value, current].map((v) => v.split(".").map(Number));
  const gt = n[0] > c[0] || (n[0] === c[0] && (n[1] > c[1] || (n[1] === c[1] && n[2] > c[2])));
  if (!gt) die(`--set ${value} is not greater than current ${current}; refusing to downgrade.`);
  if (n[0] - c[0] > 1 && !allowMajorJump) {
    die(
      `--set ${value} raises the major version from ${c[0]} to ${n[0]} (more than one ` +
        `step) — almost always a typo. If intentional, re-run with --allow-major-jump.`,
    );
  }
  return value;
}

function checkBranch() {
  const branch = get("git", ["rev-parse", "--abbrev-ref", "HEAD"]);
  if (branch !== "main") die(`Must be on main branch (currently on ${branch}).`);
}

function checkClean() {
  if (get("git", ["status", "--porcelain"])) {
    die("Working tree is not clean. Commit or stash changes first.");
  }
}

function checkInSync() {
  run("git", ["fetch", "--quiet"]);
  if (get("git", ["rev-parse", "HEAD"]) !== get("git", ["rev-parse", "origin/main"])) {
    const ahead = get("git", ["rev-list", "--count", "origin/main..HEAD"]);
    const behind = get("git", ["rev-list", "--count", "HEAD..origin/main"]);
    die(`Local main is not in sync with origin/main (${ahead} ahead, ${behind} behind). Push or pull first.`);
  }
}

function runTests() {
  console.log("Running tests (the gate CI enforces)...");
  run("npm", ["test", "--workspaces", "--if-present"]);
}

// Surface the situation the project cares about: the entviz spec / reference
// impl moving ahead of this port. Best-effort and non-fatal — it reads the
// sibling ../entviz checkout if present; absent that, it stays quiet.
function warnIfSpecBehind() {
  let ours;
  try {
    const src = readFileSync(join(REPO_ROOT, "packages/core/src/entviz.ts"), "utf8");
    ours = src.match(/SPEC_VERSION\s*=\s*"v(\d+)"/)?.[1];
  } catch {
    return;
  }
  let ref;
  try {
    const refSrc = readFileSync(join(REPO_ROOT, "../entviz/src/entviz/__init__.py"), "utf8");
    ref = refSrc.match(/SPEC_VERSION\s*=\s*"v(\d+)"/)?.[1];
  } catch {
    return; // no sibling reference checkout — nothing to compare against
  }
  if (ours === undefined || ref === undefined) return;
  if (Number(ref) > Number(ours)) {
    console.warn(
      `\n⚠  Spec drift: this port targets entviz spec v${ours}, but the reference ` +
        `impl at ../entviz is v${ref}.\n   You are releasing a port that is BEHIND the ` +
        `spec. CI's conformance job already warns about this; releasing is allowed, but\n` +
        `   make sure the version/changelog is honest about the spec level.\n`,
    );
  }
}

function setVersions(newVersion) {
  for (const path of [CORE_PKG, REACT_PKG]) {
    const pkg = readJson(path);
    pkg.version = newVersion;
    // Keep @entviz/react's pin on @entviz/core exact and in lockstep.
    if (pkg.dependencies?.[CORE_NAME]) pkg.dependencies[CORE_NAME] = newVersion;
    writeFileSync(path, JSON.stringify(pkg, null, 2) + "\n");
  }
}

function prompt(question) {
  if (!process.stdin.isTTY) {
    die(`This release requires a commit message; pass -m '<message>'.`);
  }
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => rl.question(question, (a) => { rl.close(); resolve(a.trim()); }));
}

function parseArgs(argv) {
  const opts = { part: null, explicit: null, allowMajorJump: false, message: null };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--major" || a === "--minor" || a === "--patch") {
      if (opts.part || opts.explicit) die("Pick only one of --major/--minor/--patch/--set.");
      opts.part = a.slice(2);
    } else if (a === "--set") {
      if (opts.part || opts.explicit) die("Pick only one of --major/--minor/--patch/--set.");
      opts.explicit = argv[++i] ?? die("--set requires a version argument.");
    } else if (a === "--allow-major-jump") {
      opts.allowMajorJump = true;
    } else if (a === "-m") {
      opts.message = argv[++i] ?? die("-m requires a message argument.");
    } else {
      die(`Unknown argument: ${a}`);
    }
  }
  return opts;
}

async function main() {
  const opts = parseArgs(process.argv.slice(2));
  const old = currentVersion();

  let newVersion, label;
  if (opts.explicit) {
    newVersion = parseExplicit(opts.explicit, old, opts);
    label = "set";
  } else {
    label = opts.part ?? "patch";
    newVersion = bump(old, label);
  }

  let message = opts.message;
  if (!message) {
    message = label === "patch" ? "misc fixes/enhancements" : await prompt(`Commit message for ${label} release: `);
    if (!message) die("Commit message cannot be empty.");
  }

  checkBranch();
  checkClean();
  checkInSync();
  warnIfSpecBehind();
  runTests();

  const tag = `v${newVersion}`;
  console.log(`${opts.explicit ? "Setting" : "Bumping"} ${old} -> ${newVersion}`);
  setVersions(newVersion);

  // Refresh the lockfile so its recorded versions track the manifests.
  console.log("Refreshing package-lock.json...");
  run("npm", ["install", "--package-lock-only", "--no-audit", "--no-fund"]);

  run("git", ["add", "packages/core/package.json", "packages/react/package.json", "package-lock.json"]);
  // DCO sign-off (-s): this and all repos the maintainer works in require it.
  run("git", ["commit", "-s", "-m", `Release ${tag}: ${message}`]);
  run("git", ["push", "origin", "main"]);
  run("git", ["tag", "-a", tag, "-m", `Release ${tag}: ${message}`]);
  run("git", ["push", "origin", tag]);

  console.log(`\nTagged and pushed ${tag}. The release workflow will verify, test, and publish to npm.`);
}

main().catch((e) => die(e?.stack ?? String(e)));

File: scripts/release.py
#!/usr/bin/env python3
"""Cut an entviz-js release: bump version, run the gate, commit, push, tag.

HUMAN-run by default: pushes to main and tags are reserved for a human
maintainer. An AI agent may run this script ONLY when a human has explicitly
instructed it to cut a release.

The version follows the entviz-family spec-tracking convention: 0.<spec-major>.x
means "this port is compliant with entviz spec v<spec-major>" (so 0.10.x ⇒ spec
v10, matching the Python reference's 0.10.0). A spec bump (v7 → v10) is therefore
a MINOR bump here (cut with --minor / --set); patch covers port-only changes
within a spec version. SPEC_VERSION lives in packages/core/src/entviz.ts. The
script warns if the new minor disagrees with SPEC_VERSION, and if the sibling
entviz reference (../entviz) is on a newer spec than this port claims.

Both workspace packages (@entviz/core, @entviz/react) are versioned in LOCKSTEP:
each release sets both to the same X.Y.Z and pins @entviz/react's dependency on
@entviz/core to that exact version. @entviz/core's manifest version is the
single source of truth this script reads.

Usage:
    python scripts/release.py                       # patch bump, default message
    python scripts/release.py -m "fix overlay"      # patch bump, custom message
    python scripts/release.py --minor -m "spec v10" # minor bump (e.g. spec bump)
    python scripts/release.py --major -m "1.0"      # major bump
    python scripts/release.py --set 0.10.0 -m "..." # set an explicit version
                                                    #   (must be > current; a
                                                    #    major jump > 1 needs
                                                    #    --allow-major-jump)

After the tag reaches GitHub, .github/workflows/release.yml runs the gate,
verifies the tag matches the manifest, and publishes @entviz/core to npm via
OIDC trusted publishing (no stored token).
"""

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).parent.parent
CORE_PKG = REPO_ROOT / "packages" / "core" / "package.json"
REACT_PKG = REPO_ROOT / "packages" / "react" / "package.json"
# The private playground app pins both workspace packages exactly; its pins must
# move in lockstep with a release or `npm install --package-lock-only` cannot
# resolve the just-bumped version (it is not on the registry). Its OWN version
# stays put (private, 0.0.0) — only its @entviz/* dependency pins are rewritten.
PLAYGROUND_PKG = REPO_ROOT / "apps" / "playground" / "package.json"
CORE_SRC = REPO_ROOT / "packages" / "core" / "src" / "entviz.ts"
ENTVIZ_REF = REPO_ROOT.parent / "entviz"
CORE_NAME = "@entviz/core"
REACT_NAME = "@entviz/react"


def run(cmd, *, capture=False, check=True):
    return subprocess.run(cmd, capture_output=capture, text=True, check=check, cwd=REPO_ROOT)


def get(cmd):
    return run(cmd, capture=True).stdout.strip()


def current_version():
    v = json.loads(CORE_PKG.read_text()).get("version", "")
    if not re.fullmatch(r"\d+\.\d+\.\d+", v):
        sys.exit(f"Could not read a valid X.Y.Z version from {CORE_PKG} (got {v!r}).")
    return v


def _pin_internal_deps(data, new_version):
    """Rewrite any @entviz/core or @entviz/react entry in `data.dependencies`
    to the exact new version (in place). No-op when neither is depended on."""
    deps = data.get("dependencies")
    if isinstance(deps, dict):
        for name in (CORE_NAME, REACT_NAME):
            if name in deps:
                deps[name] = new_version


def set_version(new_version):
    """Bump the two published packages in lockstep, keep every internal
    @entviz/* pin exact, and move the private playground app's pins too.

    @entviz/core and @entviz/react get the new version as their OWN version and
    have their internal pins rewritten. The playground (apps/playground) is
    private at a fixed 0.0.0, so its version is left alone, but its exact pins on
    @entviz/core/@entviz/react MUST be bumped or `npm install --package-lock-only`
    fails to resolve the just-bumped version (it is not published). json
    round-trips preserve key order and the 2-space manifest format.

    ensure_ascii=False keeps the em-dashes in descriptions literal (as
    npm/JSON.stringify wrote them), so the release diff stays version-only."""
    for path in (CORE_PKG, REACT_PKG):
        data = json.loads(path.read_text())
        data["version"] = new_version
        _pin_internal_deps(data, new_version)
        path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n")
    # Private workspace consumers: pins only, version untouched.
    for path in (PLAYGROUND_PKG,):
        data = json.loads(path.read_text())
        _pin_internal_deps(data, new_version)
        path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n")


def bump(version, part):
    major, minor, patch = (int(x) for x in version.split("."))
    if part == "major":
        return f"{major + 1}.0.0"
    if part == "minor":
        return f"{major}.{minor + 1}.0"
    return f"{major}.{minor}.{patch + 1}"


def parse_explicit_version(value, current, *, allow_major_jump=False):
    if not re.fullmatch(r"\d+\.\d+\.\d+", value):
        sys.exit(f"--set expects X.Y.Z (got {value!r}).")
    as_tuple = lambda v: tuple(int(p) for p in v.split("."))  # noqa: E731
    new, cur = as_tuple(value), as_tuple(current)
    if new <= cur:
        sys.exit(f"--set {value} is not greater than current {current}; refusing to downgrade.")
    if new[0] - cur[0] > 1 and not allow_major_jump:
        sys.exit(
            f"--set {value} raises the major version by more than one step — "
            f"almost always a typo. Re-run with --allow-major-jump if intentional."
        )
    return value


def spec_version_of(path, pattern):
    try:
        m = re.search(pattern, path.read_text())
        return m.group(1) if m else None
    except OSError:
        return None


def our_spec_major():
    return spec_version_of(CORE_SRC, r'SPEC_VERSION\s*=\s*"v(\d+)"')


def warn_if_minor_disagrees(new_version):
    """The convention ties the minor to the spec major; remind (don't block)
    if a release would break that — e.g. a spec bump cut as --patch."""
    ours = our_spec_major()
    minor = new_version.split(".")[1]
    if ours and minor != ours:
        print(
            f"\n  ⚠️  version/spec mismatch: {new_version} has minor {minor}, but "
            f"SPEC_VERSION is v{ours}.\n      The convention is 0.<spec-major>.x — a "
            f"spec bump should be a --minor (or --set) release.\n"
        )


def warn_if_behind_spec():
    """Surface the case where the reference spec is ahead of this port."""
    ours = our_spec_major()
    ref = spec_version_of(ENTVIZ_REF / "src" / "entviz" / "__init__.py",
                          r'SPEC_VERSION\s*=\s*"v(\d+)"')
    if ours and ref and int(ref) > int(ours):
        print(
            f"\n  ⚠️  spec drift: the entviz reference is on spec v{ref}, but this "
            f"port targets v{ours}.\n      Releasing now ships a port that is behind "
            f"the spec. Upgrade first, or release knowingly.\n"
        )


def check_branch():
    branch = get(["git", "rev-parse", "--abbrev-ref", "HEAD"])
    if branch != "main":
        sys.exit(f"Must be on main branch (currently on {branch!r}).")


def check_clean():
    if run(["git", "status", "--porcelain"], capture=True).stdout.strip():
        sys.exit("Working tree is not clean. Commit or stash changes first.")


def check_in_sync():
    run(["git", "fetch", "--quiet"])
    if get(["git", "rev-parse", "HEAD"]) != get(["git", "rev-parse", "origin/main"]):
        ahead = get(["git", "rev-list", "--count", "origin/main..HEAD"])
        behind = get(["git", "rev-list", "--count", "HEAD..origin/main"])
        sys.exit(
            f"Local main is not in sync with origin/main "
            f"({ahead} ahead, {behind} behind). Push or pull first."
        )


def run_gate():
    """The same gate CI enforces."""
    print("Running the gate (npm test across the workspace)...")
    run(["npm", "test", "--workspaces", "--if-present"])


def prompt_message(part):
    if not sys.stdin.isatty():
        sys.exit(f"--{part} release requires a commit message; pass -m '<message>'.")
    try:
        msg = input(f"Commit message for {part} release: ").strip()
    except (EOFError, KeyboardInterrupt):
        sys.exit("\nAborted.")
    if not msg:
        sys.exit("Commit message cannot be empty.")
    return msg


def main():
    parser = argparse.ArgumentParser(
        description="Cut a release. Defaults to --patch if no bump flag is given.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    group = parser.add_mutually_exclusive_group(required=False)
    group.add_argument("--major", dest="part", action="store_const", const="major")
    group.add_argument("--minor", dest="part", action="store_const", const="minor")
    group.add_argument("--patch", dest="part", action="store_const", const="patch")
    group.add_argument("--set", dest="explicit", metavar="X.Y.Z", default=None,
                       help="set an explicit version instead of bumping; must be > current")
    parser.add_argument("--allow-major-jump", action="store_true",
                        help="permit --set to raise the major version by more than one step")
    parser.add_argument("-m", dest="message", default=None, help="commit message")
    args = parser.parse_args()

    old = current_version()
    if args.explicit:
        new = parse_explicit_version(args.explicit, old, allow_major_jump=args.allow_major_jump)
        label = "set"
    else:
        label = args.part or "patch"
        new = bump(old, label)

    if args.message:
        message = args.message
    elif label == "patch":
        message = "misc fixes/enhancements"
    else:
        message = prompt_message(label)

    check_branch()
    check_clean()
    check_in_sync()
    warn_if_minor_disagrees(new)
    warn_if_behind_spec()
    run_gate()

    tag = f"v{new}"
    verb = "Setting" if args.explicit else "Bumping"
    print(f"{verb} {old} -> {new}")
    set_version(new)
    # Refresh package-lock.json so its recorded versions track the manifests
    # (does not upgrade dependencies).
    run(["npm", "install", "--package-lock-only", "--no-audit", "--no-fund"])

    run(["git", "add",
         "packages/core/package.json", "packages/react/package.json",
         "apps/playground/package.json", "package-lock.json"])
    run(["git", "commit", "-s", "-m", f"Release {tag}: {message}"])
    run(["git", "push", "origin", "main"])
    run(["git", "tag", "-a", tag, "-m", f"Release {tag}: {message}"])
    run(["git", "push", "origin", tag])

    print(f"Tagged and pushed {tag}. The release workflow will gate + publish to npm (OIDC).")


if __name__ == "__main__":
    main()
